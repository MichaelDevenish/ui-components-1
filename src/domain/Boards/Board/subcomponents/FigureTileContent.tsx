import React from 'react'

import { Props, Variables } from '../../../../common'
import {
  ButtonDescriptionText,
  ButtonTextContentWrapper,
  ButtonTitleText,
  ButtonWrapper,
  CenteredContentWrapper,
  ChildrenWrapper,
  HeadingLine,
  HeadingWrapper,
  StyleFontAwesomeIcon,
  StyledFigureText,
  StyledHeadingText,
  StyledIntelliIcon,
  StyledTileLabel,
  TileContentWrapper
} from './style'

interface IBoardFigureTileContentProps {
  /** Text displayed above the content of tile */
  label?: string
  /** Text displayed at the top of tile content */
  heading?: string
  /** Text displayed in the heading */
  headingFigure?: string
  /** Text displayed after heading */
  subheading?: string
  /** Text displayed in the subheading */
  subheadingFigure?: string
  /** If true, displays heading in a warning style */
  isHeadingWarning?: boolean
  /** If true, displays subheading in a warning style */
  isSubheadingWarning?: boolean
  /** If yes the tile content would has a margin to ensure the hover label does not overlap with the content */
  hasHoverMargin?: boolean
}

class FigureTileContent extends React.PureComponent<IBoardFigureTileContentProps, never> {
  public static defaultProps: Partial<IBoardFigureTileContentProps> = {
    isHeadingWarning: false,
    isSubheadingWarning: false
  }

  private get headingFigure (): JSX.Element | null {
    const {
      headingFigure,
      isHeadingWarning
    } = this.props

    if (headingFigure) {
      return (
        <StyledFigureText
          type={Props.TypographyType.DisplayLarge}
          color={isHeadingWarning ? Variables.Color.r400 : Variables.Color.n800}
          isUpper
          textWidth={this.figureTextWidth}
        >
          {headingFigure}
        </StyledFigureText>
      )
    }

    return null
  }

  private get heading (): JSX.Element | null {
    const {
      heading,
      isHeadingWarning
    } = this.props

    if (heading) {
      return (
        <StyledHeadingText
          type={Props.TypographyType.Heading}
          color={isHeadingWarning ? Variables.Color.r400 : Variables.Color.n800}
          isUpper
        >
          {heading}
        </StyledHeadingText>
      )
    }

    return null
  }

  private get subheadingFigure (): JSX.Element | null {
    const {
      subheadingFigure,
      isSubheadingWarning
    } = this.props

    if (subheadingFigure) {
      return (
        <StyledFigureText
          type={Props.TypographyType.Display}
          color={isSubheadingWarning ? Variables.Color.r400 : Variables.Color.n800}
          isUpper
          textWidth={this.figureTextWidth}
        >
          {subheadingFigure}
        </StyledFigureText>
      )
    }

    return null
  }

  private get subheading (): JSX.Element | null {
    const {
      subheading,
      isSubheadingWarning
    } = this.props

    if (subheading) {
      return (
        <StyledHeadingText
          type={Props.TypographyType.Small}
          isUpper
          color={isSubheadingWarning ? Variables.Color.r400 : Variables.Color.n800}
        >
          {subheading}
        </StyledHeadingText>
      )
    }

    return null
  }

  private get figureTextWidth (): number {
    const {
      headingFigure,
      subheadingFigure
    } = this.props

    const headingFigureTextLength = headingFigure ? headingFigure.length : 0
    const subheadingFigureTextLength = subheadingFigure ? subheadingFigure.length : 0

    return Math.max(headingFigureTextLength * 20, subheadingFigureTextLength * 15) + 16
  }

  private get headings (): JSX.Element | null {
    const {
      heading,
      subheading
    } = this.props

    if (heading || subheading) {
      return (
        <HeadingWrapper>
          <HeadingLine>
            {this.headingFigure}
            {this.heading}
          </HeadingLine>
          <HeadingLine>
            {this.subheadingFigure}
            {this.subheading}
          </HeadingLine>
        </HeadingWrapper>
      )
    }

    return null
  }

  private get content (): JSX.Element {
    const {
      children,
      hasHoverMargin
    } = this.props

    return (
      <ChildrenWrapper
        hasHoverMargin={hasHoverMargin}
      >
        {children}
      </ChildrenWrapper>
    )
  }

  private get tileLabel (): JSX.Element | null {
    const {
      label
    } = this.props

    if (label) {
      return (
        <StyledTileLabel>
          {label}
        </StyledTileLabel>
      )
    }

    return null
  }

  public render (): JSX.Element {
    return (
      <TileContentWrapper
      >
        {this.tileLabel}
        {this.headings}
        {this.content}
      </TileContentWrapper>
    )
  }
}

export {
  FigureTileContent
}
