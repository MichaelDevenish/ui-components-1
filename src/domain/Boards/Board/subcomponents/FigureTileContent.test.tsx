import { shallow } from 'enzyme'
import React from 'react'

import { FigureTileContent } from './FigureTileContent'

describe('<FigureTileContent />', () => {
  it(`should render a figure tile content with nothing`, () => {
    const wrapper = shallow(
      <FigureTileContent  />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a figure tile content with label`, () => {
    const wrapper = shallow(
      <FigureTileContent
        label='dummy label'
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a figure tile content with normal headings`, () => {
    const wrapper = shallow(
      <FigureTileContent
        heading='dummy heading'
        headingFigure='1'
        subheading='dummy subheading'
        subheadingFigure='35'
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a figure tile content with warning headings`, () => {
    const wrapper = shallow(
      <FigureTileContent
        heading='dummy heading'
        headingFigure='1'
        subheading='dummy subheading'
        subheadingFigure='35'
        isHeadingWarning
        isSubheadingWarning
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a figure tile content with heading only`, () => {
    const wrapper = shallow(
      <FigureTileContent
        heading='dummy heading'
        headingFigure='1'
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it(`should render a figure tile content with subheading only`, () => {
    const wrapper = shallow(
      <FigureTileContent
        subheading='dummy subheading'
        subheadingFigure='35'
      />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
