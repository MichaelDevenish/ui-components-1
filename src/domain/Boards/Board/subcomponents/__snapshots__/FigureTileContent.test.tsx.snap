// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FigureTileContent /> should render a figure tile content with heading only 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        textWidth={36}
        type="display-large"
      >
        1
      </Styled(Text)>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        type="heading"
      >
        dummy heading
      </Styled(Text)>
    </styled.div>
    <styled.div />
  </styled.div>
  <styled.div />
</styled.div>
`;

exports[`<FigureTileContent /> should render a figure tile content with label 1`] = `
<styled.div>
  <styled.label>
    dummy label
  </styled.label>
  <styled.div />
</styled.div>
`;

exports[`<FigureTileContent /> should render a figure tile content with normal headings 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        textWidth={46}
        type="display-large"
      >
        1
      </Styled(Text)>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        type="heading"
      >
        dummy heading
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        textWidth={46}
        type="display"
      >
        35
      </Styled(Text)>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        type="small"
      >
        dummy subheading
      </Styled(Text)>
    </styled.div>
  </styled.div>
  <styled.div />
</styled.div>
`;

exports[`<FigureTileContent /> should render a figure tile content with nothing 1`] = `
<styled.div>
  <styled.div />
</styled.div>
`;

exports[`<FigureTileContent /> should render a figure tile content with subheading only 1`] = `
<styled.div>
  <styled.div>
    <styled.div />
    <styled.div>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        textWidth={46}
        type="display"
      >
        35
      </Styled(Text)>
      <Styled(Text)
        color="#262D33"
        isUpper={true}
        type="small"
      >
        dummy subheading
      </Styled(Text)>
    </styled.div>
  </styled.div>
  <styled.div />
</styled.div>
`;

exports[`<FigureTileContent /> should render a figure tile content with warning headings 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <Styled(Text)
        color="#F25749"
        isUpper={true}
        textWidth={46}
        type="display-large"
      >
        1
      </Styled(Text)>
      <Styled(Text)
        color="#F25749"
        isUpper={true}
        type="heading"
      >
        dummy heading
      </Styled(Text)>
    </styled.div>
    <styled.div>
      <Styled(Text)
        color="#F25749"
        isUpper={true}
        textWidth={46}
        type="display"
      >
        35
      </Styled(Text)>
      <Styled(Text)
        color="#F25749"
        isUpper={true}
        type="small"
      >
        dummy subheading
      </Styled(Text)>
    </styled.div>
  </styled.div>
  <styled.div />
</styled.div>
`;
